src/parser.rs

```diff
@@ -269,15 +270,56 @@ impl Library {
                         "virtual-method" => try!(ignore_element(parser)),
                         "doc" => doc = try!(read_text(parser)),
                         "union" => {
-                            use self::Type::*;
-                            if let Union(u) = try!(self.read_union(parser, ns_id, attrs)) {
+                            if let Type::Union(mut u) = try!(self.read_union(
+                                parser,
+                                ns_id,
+                                &attributes,
+                                Some(&class_name),
+                                Some(&c_type)
+                            )) {
+                                let mut field_name = String::new();
+                                for attr in &attributes {
+                                    match attr.name.local_name.as_ref() {
+                                        "name" => {
+                                            field_name = attr.value.clone();
+                                            break;
+                                        }
+                                        _ => {}
+                                    }
+                                }
+
+                                if field_name == "" {
+                                    field_name = format!("u{}", union_count);
+                                }
+
+                                u = Union {
+                                    name: format!("{}_{}", class_name, field_name),
+                                    c_type: Some(format!("{}_{}", c_type, field_name)),
+                                    ..u
+                                };
+
                                 let u_doc = u.doc.clone();
-                                let type_id = Type::union(self, u, ns_id);
+                                let ctype = u.c_type.clone();
+
+                                let type_id = {
+                                    #[cfg(not(feature = "use_unions"))]
+                                    {
+                                        Type::union(self, u, INTERNAL_NAMESPACE)
+                                    }
+                                    #[cfg(feature = "use_unions")]
+                                    {
+                                        Type::union(self, u, ns_id)
+                                    }
```

Comment:
You right, I mixed up `INTERNAL_NAMESPACE` and `MAIN_NAMESPACE`.
All 3 of these better remain